File: android/src/main/kotlin/com/example/platform_bridge/PlatformBridgePlugin.kt
package com.example.platform_bridge

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.*
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** 
 * PlatformBridgePlugin - 用于Flutter与原生Android平台间双向通信的插件
 * 
 * 功能包括：
 * 1. Flutter向原生发送数据（sendToNative）
 * 2. 原生向Flutter发送数据（sendDataToFlutter）
 * 3. 原生监听来自Flutter的数据（listenFromFlutter）
 */
class PlatformBridgePlugin : FlutterPlugin, MethodCallHandler {
    /// 用于与Flutter通信的方法通道
    private lateinit var channel : MethodChannel
    
    /// 存储监听器的map，用于监听来自Flutter的数据
    private val flutterListeners = mutableMapOf<String, (Any?) -> Unit>()
    
    companion object {
        /// 通道名称
        private const val CHANNEL_NAME = "com.seven.platform_bridge/platform_bridge"
        
        /// 获取平台版本号的方法名
        private const val METHOD_GET_PLATFORM_VERSION = "getPlatformVersion"
        
        /// Flutter向原生发送数据的方法名
        private const val METHOD_SEND_TO_NATIVE = "sendToNative"
        
        /// 原生向Flutter发送数据的方法名
        private const val METHOD_SEND_TO_FLUTTER = "sendToFlutter"
        
        /// 参数名称
        private const val PARAM_NAME = "name"
        
        /// 参数数据
        private const val PARAM_DATA = "data"
        
        /// 单例实例
        @Volatile
        private var INSTANCE: PlatformBridgePlugin? = null

        /**
         * 获取插件单例实例
         */
        @JvmStatic
        fun getInstance(): PlatformBridgePlugin {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: PlatformBridgePlugin().also { INSTANCE = it }
            }
        }
    }

    /// 默认构造函数，供Flutter框架使用
    constructor()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        LogUtils.d("Plugin attached to engine")
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
        
        // 保存单例引用
        INSTANCE = this
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        LogUtils.d("Received method call: ${call.method}")
        
        when (call.method) {
            METHOD_GET_PLATFORM_VERSION -> {
                LogUtils.d("Getting platform version")
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            METHOD_SEND_TO_NATIVE -> {
                LogUtils.d("Handling sendToNative method call")
                
                val name = call.argument<String>(PARAM_NAME)
                val data = call.argument<Any>(PARAM_DATA)
                
                if (name != null) {
                    LogUtils.d("Received data from Flutter: name=$name, data=$data")
                    
                    // 处理从Flutter发送过来的数据
                    handleDataFromFlutter(name, data)
                    
                    result.success(null)
                } else {
                    LogUtils.e("Name parameter is null")
                    result.error("INVALID_ARGUMENT", "Name cannot be null", null)
                }
            }
            else -> {
                LogUtils.w("Method not implemented: ${call.method}")
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        LogUtils.d("Plugin detached from engine")
        channel.setMethodCallHandler(null)
        
        // 清除单例引用
        INSTANCE = null
    }
    
    /**
     * 处理由Flutter发送过来的数据
     * 
     * @param name 数据标识符
     * @param data 要处理的数据
     */
    private fun handleDataFromFlutter(name: String, data: Any?) {
        LogUtils.d("Handling data from Flutter: name=$name, data=$data")
        
        // 检查实例特定的监听器
        flutterListeners[name]?.let { listener ->
            LogUtils.d("Calling instance listener for name: $name")
            listener(data)
        }
    }
    
    /**
     * 向Flutter发送数据
     * 
     * @param name 数据标识符
     * @param data 要发送的数据
     */
    fun sendDataToFlutter(name: String, data: Any?) {
        LogUtils.d("Sending data to Flutter: name=$name, data=$data")
        
        // 使用MethodChannel向Flutter发送数据
        val args = hashMapOf<String, Any?>(
            PARAM_NAME to name,
            PARAM_DATA to data
        )
        channel.invokeMethod(METHOD_SEND_TO_FLUTTER, args)
    }
    
    /**
     * 监听来自Flutter的数据
     * 
     * @param name 数据标识符
     * @param listener 接收数据的回调函数
     */
    fun listenFromFlutter(name: String, listener: (Any?) -> Unit) {
        LogUtils.d("Setting up listener for name: $name")
        flutterListeners[name] = listener
    }
}

File: ios/Classes/PlatformBridgePlugin.swift
import Flutter
import UIKit

/**
 * PlatformBridgePlugin - 用于Flutter与原生iOS平台间双向通信的插件
 *
 * 功能包括：
 * 1. Flutter向原生发送数据（sendToNative）
 * 2. 原生向Flutter发送数据（sendDataToFlutter）
 * 3. 原生监听来自Flutter的数据（listenFromFlutter）
 */
public class PlatformBridgePlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    
    // 存储监听器的字典，用于监听来自Flutter的数据
    private var flutterListeners: [String: (Any?) -> Void] = [:]
    
    // MARK: - Constants
    private static let CHANNEL_NAME = "com.seven.platform_bridge/platform_bridge"
    private static let METHOD_GET_PLATFORM_VERSION = "getPlatformVersion"
    private static let METHOD_SEND_TO_NATIVE = "sendToNative"
    private static let METHOD_SEND_TO_FLUTTER = "sendToFlutter"
    private static let PARAM_NAME = "name"
    private static let PARAM_DATA = "data"
    
    // MARK: - Singleton instance
    private static var sharedInstance: PlatformBridgePlugin?
    
    // 私有初始化方法
    private override init() {
        super.init()
        
        LogUtils.info("PlatformBridgePlugin initialized")
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: PlatformBridgePlugin.CHANNEL_NAME, binaryMessenger: registrar.messenger())
        let instance = PlatformBridgePlugin.getInstance()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        LogUtils.info("PlatformBridgePlugin registered with registrar")
    }
    
    /**
     * 获取插件单例实例
     */
    public static func getInstance() -> PlatformBridgePlugin {
        if let instance = sharedInstance {
            return instance
        }
        let instance = PlatformBridgePlugin()
        // 保存单例引用
        PlatformBridgePlugin.sharedInstance = instance
        return instance
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        LogUtils.debug("Received method call: \(call.method)")
        
        switch call.method {
        case PlatformBridgePlugin.METHOD_GET_PLATFORM_VERSION:
            LogUtils.info("Getting platform version")
            result("iOS " + UIDevice.current.systemVersion)
        case PlatformBridgePlugin.METHOD_SEND_TO_NATIVE:
            guard let arguments = call.arguments as? [String: Any],
                  let name = arguments[PlatformBridgePlugin.PARAM_NAME] as? String else {
                LogUtils.error("Invalid arguments: Missing name or data")
                
                result(FlutterError(code: "INVALID_ARGUMENT",
                                   message: "Missing name or data",
                                   details: nil))
                return
            }
            
            let data = arguments[PlatformBridgePlugin.PARAM_DATA]
            
            LogUtils.debug("Received data from Flutter: name=\(name), data=\(String(describing: data))")
            
            // 处理由Flutter发送过来的数据
            handleDataFromFlutter(name: name, data: data)
            
            result(nil)
        default:
            LogUtils.warning("Method not implemented: \(call.method)")
            result(FlutterMethodNotImplemented)
        }
    }
    
    /**
     * 处理由Flutter发送过来的数据
     *
     * - Parameters:
     *   - name: 数据标识符
     *   - data: 要处理的数据
     */
    private func handleDataFromFlutter(name: String, data: Any?) {
        LogUtils.debug("Handling data from Flutter: name=\(name), data=\(String(describing: data))")
        
        // 检查实例特定的监听器
        if let listener = flutterListeners[name] {
            LogUtils.debug("Calling instance listener for name: \(name)")
            listener(data)
        }
    }
    
    /**
     * 向Flutter发送数据
     *
     * - Parameters:
     *   - name: 数据标识符
     *   - data: 要发送的数据
     */
    public func sendDataToFlutter(name: String, data: Any?) {
        LogUtils.debug("Sending data to Flutter: name=\(name), data=\(String(describing: data))")
        
        let args: [String: Any?] = [
            PlatformBridgePlugin.PARAM_NAME: name,
            PlatformBridgePlugin.PARAM_DATA: data
        ]
        channel?.invokeMethod(PlatformBridgePlugin.METHOD_SEND_TO_FLUTTER, arguments: args)
    }
    
    /**
     * 监听来自Flutter的数据
     *
     * - Parameters:
     *   - name: 数据标识符
     *   - listener: 接收数据的回调函数
     */
    public func listenFromFlutter(name: String, listener: @escaping (Any?) -> Void) {
        LogUtils.debug("Setting up listener for Flutter data: name=\(name)")
        flutterListeners[name] = listener
    }
}